homeassistant/components/meater/sensor.py

```diff
@@ -0,0 +1,112 @@
+"""The Meater Temperature Probe integration."""
+from enum import Enum
+
+from homeassistant.components.sensor import SensorDeviceClass, SensorEntity
+from homeassistant.const import TEMP_CELSIUS
+from homeassistant.core import callback
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import DOMAIN
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the entry."""
+    coordinator: DataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id][
+        "coordinator"
+    ]
+
+    @callback
+    def async_update_data():
+        """Handle updated data from the API endpoint."""
+        if not coordinator.last_update_success:
+            return
+
+        devices = coordinator.data
+        entities = []
+        known_probes: set = hass.data[DOMAIN]["known_probes"]
+
+        # Add entities for temperature probes which we've not yet seen
+        for dev in devices:
+            if dev.id in known_probes:
+                continue
+
+            entities.append(
+                MeaterProbeTemperature(
+                    coordinator, dev.id, TemperatureMeasurement.Internal
+                )
+            )
+            entities.append(
+                MeaterProbeTemperature(
+                    coordinator, dev.id, TemperatureMeasurement.Ambient
+                )
+            )
+            known_probes.add(dev.id)
+
+        async_add_entities(entities)
+
+        return devices
+
+    # Add a subscriber to the coordinator to discover new temperature probes
+    coordinator.async_add_listener(async_update_data)
```

Comment:
We should unsubscribe this listener when unloading the config entry. We can use `entry.async_on_unload`.